### PR TITLE
Generic adapters implementation

### LINKED ISSUE #2526: Adapters + LLama -- re-design.
### Describe the bug

It's not a bug, just a discussion. I think that people including @Adel-Moumen @poonehmousavi @mravanelli and maybe @pplantinga @asumagic may want to engage. 

Adapters, or, more generally, altering an existing pre-trained model (you can see it as an object originating from the Pretrainer or checkpointer) is something becoming more and more common. Due to this, we, imho, **must** define a proper design in SpeechBrain to do so. Recently, I implemented LoRA and Houlsby for our Transformer, on my side. But I also realised that @poonehmousavi did some work for Llama [here](https://github.com/speechbrain/speechbrain/blob/b12b6837a23590fb267b6adcd1a0640f5a1070ca/speechbrain/lobes/models/huggingface_transformers/llama2.py#L220). I don't think we are doing this correctly. The code in Llama 2, for instance, might be hard to understand, and some functions (like the one replacing a module in an existing module) should be generalised and considered a general SpeechBrain util. My strategy would be to create an Adapters.py in lobes where we could put everything relating to them, instead of having them randomly appearing in lobes files. 

What do you folks think? 

### Expected behaviour

Respect the Zen of SpeechBrain. 

### To Reproduce

_No response_

### Environment Details

_No response_

### Relevant Log Output

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by pplantinga**
Closes #2526, #2534

Here is a proposal for how we can add adapters (including LoRA) to the toolkit. This branch is based on #2534 - and it also implements flexible layer selection and small checkpoints. 

There's a few more things that would be nice to have but I personally don't think they're necessary before merge.
* a `merge_and_unload()` type function for LoRA-type layers that reintegrates the adapter weights to the original model
* the capability to use adapters from `peft` library -- they have an extensive collection that will probably update regularly
* more adapter types

If anyone thinks these are urgent we can work on adding them to this PR.

UPDATE (see below): This works with PEFT now.

### CODE REVIEW

**Comment by TParcollet** [comment]
Hey Peter, I would think that using peft could be a nice add to this PR. Not critical, but a nice add! I will have a look at the code, thanks:

File: recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+whisper_pretrained: !new:speechbrain.lobes.models.huggingface_transformers.whisper.Whisper
+    source: !ref <whisper_hub>
+    freeze_encoder: !ref <freeze_encoder>
+    save_path: !ref <whisper_folder>
+    language: "english"
+    task: "transcribe"
+    sampling_rate: !ref <sampling_rate>
+
+whisper: !new:speechbrain.nnet.adapters.AdaptedModel
+    model_to_adapt: !ref <whisper_pretrained>
```
Comment (TParcollet) [inline recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml:124]: Can we select only sub blocks here? Like say, only Transformer.encoder[5] for instance?

File: recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    adapter_class: !name:speechbrain.nnet.adapters.LoRA
+    rank: !ref <lora_rank>
+    target_layers: [all-linear]
```
Comment (TParcollet) [inline recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml:127]: We will need to be carreful about such a naming for a variable as it does not explicitely target elements, it could become hard to follow (if we want to target a single layer in the network?

**Comment by pplantinga** [comment]
I tried this recipe with a `peft` layer and it *just worked* to my amazement. Here's the exact change I made:

File: recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml (hunk 2 above)
Comment (TParcollet) [reply recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml:127]: I don't know what would be the best tbf, hum... I like the boolean approach I think.

**Comment by TParcollet** [comment]
@pplantinga are the checkpointing feature working as well? We should make sure it works with Pretrainer also, not just checkpointing  I blieve.

File: recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml (hunk 2 above)
Comment (pplantinga) [reply recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml:127]: It is on the same level right now. Another option would be instead of `**kwargs` add an `adapter_kwargs` argument with dictionary params to pass to the adapter layer.

Comment (pplantinga) [reply recipes/LibriSpeech/ASR/transformer/hparams/train_whisper_lora.yaml:127]: Like:

```yaml
whisper: !new:speechbrain.nnet.adapters.AdaptedModel
  model_to_adapt: !ref <whisper_pretrained>
  adapter_class: !name:speechbrain.nnet.adapters.LoRA
  all_linear: True
  adapter_kwargs:
    rank: !ref <lora_rank>
    scale: !ref <lora_scale>
```

**Comment by TParcollet** [comment]
@Adel-Moumen I think we will want this in v1.0.1 And it looks ready to me?

**Comment by TParcollet** [comment]
@poonehmousavi could you review and test the code as mentioned? It looks ready to me. Thanks!

**Comment by poonehmousavi** [comment]
> @poonehmousavi could you review and test the code as mentioned? It looks ready to me. Thanks!

Sure. I will do it by tomorrow.

**Comment by poonehmousavi** [comment]
@pplantinga have you tested it with pretrainer using for interfaces?  also have you checked how it works with quantization?(like QLORA)

**Comment by pplantinga** [comment]
> @pplantinga have you tested it with pretrainer using for interfaces?
I tested this and it worked, but had warnings due to loading only trained params. I have fixed this now.

The yaml I used is here:

```yaml
whisper_hub: openai/whisper-small.en
lora_rank: 16
language: "english"
sample_rate: 16000

min_decode_ratio: 0.0
max_decode_ratio: 1.0
test_beam_size: 8

whisper_pretrained: !new:speechbrain.lobes.models.huggingface_transformers.whisper.Whisper
    source: !ref <whisper_hub>
    save_path: .
    language: !ref <language>
    task: "transcribe"
    sampling_rate: !ref <sample_rate>

whisper: !new:speechbrain.nnet.adapters.AdaptedModel
    model_to_adapt: !ref <whisper_pretrained>
    adapter_class: !name:speechbrain.nnet.adapters.LoRA
    all_linear: True
    adapter_kwargs:
        rank: !ref <lora_rank>

test_search: !new:speechbrain.decoders.seq2seq.S2SWhisperBeamSearcher
    module: [!ref <whisper>]
    min_decode_ratio: !ref <min_decode_ratio>
    max_decode_ratio: !ref <max_decode_ratio>
    beam_size: !ref <test_beam_size>

modules:
    whisper: !ref <whisper>
    decoder: !ref <test_search>

pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
    loadables:
        whisper: !ref <whisper>
```

And python:
```python
model = sb.inference.ASR.WhisperASR.from_hparams(".", "lora_pre.yaml", savedir="results/whisper/1987/save/CKPT+2024-06-05+18-30-33+00")
model.transcribe_file("speechbrain/asr-streaming-conformer-librispeech/test-en.wav")
```

> also have you checked how it works with quantization?(like QLORA)

I am not very familiar with QLoRA, it seems there's additional setup needed to get this to work.

**Comment by pplantinga** [comment]
One epoch (100h) results for Whisper Small.en, published results are `test-clean=3.05` and `test-other=7.53`:

```
speechbrain.utils.train_logger - Epoch loaded: 1 - test loss: 9.73e-01, test CER: 1.03, test WER: 2.81
speechbrain.utils.train_logger - Epoch loaded: 1 - test loss: 9.86e-01, test CER: 1.08, test WER: 2.90
speechbrain.utils.train_logger - Epoch loaded: 1 - test loss: 1.22, test CER: 3.00, test WER: 6.57
```

**Comment by TParcollet** [comment]
@pplantinga Should we merge this maybe?

**Comment by pplantinga** [comment]
There are more features that could be added but I think this is ready for merge as-is and the rest can be added later.

**Review by TParcollet — APPROVED** [review]
LGTM now :)
